streams/src/test/java/org/apache/kafka/streams/kstream/RepartitionTopicNamingTest.java

```diff
@@ -0,0 +1,579 @@
+
+        assertThat(optimizedTopology, is(EXPECTED_OPTIMIZED_TOPOLOGY));
+        // only one repartition topic
+        assertThat(1, is(getCountOfRepartitionTopicsFound(optimizedTopology, repartitionTopicPattern)));
+        // the first named repartition topic
+        assertTrue(optimizedTopology.contains(firstRepartitionTopicName + "-repartition"));
+
+
+        assertThat(unOptimizedTopology, is(EXPECTED_UNOPTIMIZED_TOPOLOGY));
+        // now 4 repartition topic
+        assertThat(4, is(getCountOfRepartitionTopicsFound(unOptimizedTopology, repartitionTopicPattern)));
+        // all 4 named repartition topics present
+        assertTrue(unOptimizedTopology.contains(firstRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(secondRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(thirdRepartitionTopicName + "-repartition"));
+        assertTrue(unOptimizedTopology.contains(fourthRepartitionTopicName + "-left-repartition"));
+
+    }
+
+    // can't use same repartition topic name
+    @Test(expected = TopologyException.class)
+    public void shouldFailWithSameRepartitionTopicName() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        builder.<String, String>stream("topic").selectKey((k, v) -> k).groupByKey(Grouped.named("grouping")).count().toStream();
+        builder.<String, String>stream("topicII").selectKey((k, v) -> k).groupByKey(Grouped.named("grouping")).count().toStream();
+        builder.build();
+    }
+
+    // each KGroupedStream will result in repartition, can' reuse
+    // KGroupedStreams when specifying repartition topic names
+    // need to have separate groupByKey calls when naming repartition topics
+    // see test shouldHandleUniqueGroupedInstances below for an example
+    @Test(expected = TopologyException.class)
+    public void shouldFailWithSameRepartitionTopicNameUsingSameKGroupedStream() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        final KGroupedStream<String, String> kGroupedStream = builder.<String, String>stream("topic").selectKey((k, v) -> k).groupByKey(Grouped.named("grouping"));
+        kGroupedStream.windowedBy(TimeWindows.of(10)).count();
+        kGroupedStream.windowedBy(TimeWindows.of(30)).count();
+        builder.build();
+    }
+
+
+    @Test
+    public void shouldHandleUniqueGroupedInstances() {
+        final StreamsBuilder builder = new StreamsBuilder();
+        final KStream<String, String> kStreamKeySelect = builder.<String, String>stream("topic").selectKey((k, v) -> k);
+        kStreamKeySelect.groupByKey(Grouped.named("grouping")).windowedBy(TimeWindows.of(10)).count();
+        kStreamKeySelect.groupByKey(Grouped.named("groupingII")).windowedBy(TimeWindows.of(30)).count();
+        builder.build();
+    }
+
+
+
+    @Test
+    public void shouldKeepRepartitionTopicNameForJoins() {
```

Comment:
Seems for Joined we do not have a test to check for naming uniqueness yet, could we add one?

Reply:
ack